### ISSUE TITLE
Ability to work with generic types in ProtoFlux.

### ISSUE THREAD

**Issue by JackTheFoxOtter** [Yellow-Dog-Man/Resonite-Issues#3241]
### Is your feature request related to a problem? Please describe.

This is a thing I've come across a couple of times now, and I think it would be a nice addition to ProtoFlux to simplify / generalize some setups. Note that I'm not quite sure how a solution like this could look like, but I do want to explain with an example:

I'm writing a piece of code that operates on an `IField<T>` input object. After the operation, it is supposed to write the `IField<T>` into a dynamic variable. The problem I'm having is that I presently don't have the ability to write this code in a generic way. I can treat the input object as an `IField` reference by casting the input type (ex: `IField<string>` -> `IField`), but that's where the generalization options end. Currently to call `WriteDynamicObjectVariable<T>`, I need to choose a type when spawning the node. (So for example `WriteDynamicObjectVariable<IField<string>>`). This means that I now need to check my input type, figure out what generic type it corresponds to, and call the correct `WriteDynamicObjectVariable<T>` node for that type, instead of being able to write the code in a generic way. This requires me to hard-code in what types my code support, which is pretty ugly and laborous.

### Describe the solution you'd like

I'd like to be able to work with generic types in ProtoFlux. I want to be able to write code that calls a generic node in a generic way.

### Describe alternatives you've considered

Hard-coding in supported types, but as pointed out, this is pretty limiting.

**Comment by JackTheFoxOtter** [comment]
Boxing / unboxing doesn't change anything here. You cannot use generic nodes without specifying an explicit type to use presently.

**Comment by JackTheFoxOtter** [comment]
There's another use-case for this. Once we get nested nodes (#564), it would be extremely useful to be able to specify our own custom generic nodes, for example math operations that work on all number types. Otherwise you'd be required to create multiple versions of the same node for multiple number types, which would quickly become a lot of work, especially if you need to then make a change to them.

**Comment by Nytra** [comment]
> Boxing / unboxing doesn't change anything here. You cannot use generic nodes without specifying an explicit type to use presently.

You can use DynamicImpulseTriggerWithObject<object> and then determine what to do by checking the type

**Comment by JackTheFoxOtter** [comment]
That is functionally identical as casting the input to `IField` in my example. It doesn't change the fact that I then would have to check the type and multiplex the flow to multiple different generic nodes, like `WriteDynamicObjectVariable<string>`, `WriteDynamicObjectVariable<float>`, ...

**Comment by JackTheFoxOtter** [comment]
> That is functionally identical to casting the input to `IField` in my example. It doesn't change the fact that I then would have to check the type and multiplex the flow to multiple different generic nodes, like `WriteDynamicObjectVariable<string>`, `WriteDynamicObjectVariable<float>`, ...

To add even more context, this is how I currently have to write the code for my use-case:
![Image](https://github.com/user-attachments/assets/572ec4a8-e77f-4ebb-8c4f-fd7f9b15e35c)

I would like to be able to treat the input type (in this case my `IField`) as a generic `IField<T>` type and just directly call `WriteDynamicObjectVariable<T>` without specifying that `<T>` in code. So regardless of whether my input is an `IField<string>`, `IField<float>` or `IField<Chirality>`, the code would work generally without needing to write out explicit handling for all of those types.

**Comment by 989onan** [comment]
Basically, please Frooxius add Dummy back and my life is yours

**Comment by JackTheFoxOtter** [comment]
No, this is not how the "dummy" type worked. Also generics are more complicated than that. Generics can be "limited" to only certain types, like only numeric or only data model types. I'd imagine that would have to be respected here. That's why I mentioned that I don't fully know how a proper solution for this would look like. But I'm pretty sure it wouldn't be "just add the Dummy type back".

I just know that especially once we get nested nodes this would be really powerful for writing more generic and reusable code.

**Comment by mpmxyz** [comment]
#4621 was closed in favor of this ticket. This emphasizes the requirement to also be able to pick a type during ProtoFlux execution. (what you would do with reflection in C#)

**Comment by charlie-sans** [comment]
i have partially solved this issue. using execution contexts could be used to evaluate custom class start/end aswell as method start and end nodes and what not. the idea is to use these to build what pretty much is just a CIL or a IR representation. which then can be used to evaluate with custom class contexts. 

of course this is only what i found and may be different depending on the required implementation details for rsonite.
